Create tombstones for forms and cases on hard deletion

minhaminha opened the pull request:
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

[Jira ticket](https://dimagi.atlassian.net/browse/SAAS-15221)
This PR adds an additional step in the daily eligible data deletion task. On top of hard deleting eligible forms and cases, it will create a new `DeletedSQLDoc` object for each object that will serve as a long term record of its deletion. As of now, the `DeletedSQLDoc` objects for a particular domain will persist until the domain itself and all its data is deleted.  

This PR also increases amount of time a soft deleted object needs to wait before being hard deleted from 30 to 90 days.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

Tested locally. This will hard delete form and case data that's been soft deleted over 90 days ago. 

Before merging/deploying to production, I'll manually run `permanently_delete_and_tombstone_eligible_data` on a private release on both staging and production to apply this change to existing forms/cases eligible for deletion while monitoring datadog for any performance hits. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

Added more tests to cover tombstone creation logic.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

n/a

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

minhaminha committed:
Enable tombstone creation for automated hard deletion task + add tests

minhaminha committed:
Override form and case delete to require tombstones, except for tests, domain deletion and system cases/forms

millerdev commented on corehq/form_processor/models/cases.py:
```diff
@@ -834,6 +835,16 @@ def save(self, *args, with_tracked_models=False, **kw):
+                'Cannot delete case without leaving a tombstone except during testing, domain deletion or '
+                'when deleting system cases')
+        if leave_tombstone:
+            case = self.get(self._id)
+            create_deleted_sql_doc(case.case_id, 'form_processor.CommCareCase', case.domain, case.deleted_on)
```
```suggestion
            create_deleted_sql_doc(self.case_id, 'form_processor.CommCareCase', self.domain, self.deleted_on)
```

millerdev commented on corehq/form_processor/models/cases.py:
```diff
@@ -834,6 +835,16 @@ def save(self, *args, with_tracked_models=False, **kw):
         except DatabaseError as e:
             raise CaseSaveError(e)
 
+    def delete(self, leave_tombstone=False):
```
Would be great to have a test that deletes a case using a tombstone.

millerdev replied:
Would it make sense to have the default be `leave_tombstone=True`? It seems like we would want that most of the time, excluding tests, which we can handle with a different workaround.

millerdev commented on corehq/form_processor/models/forms.py:
```diff
@@ -723,6 +724,16 @@ def unarchive(self, user_id=None, trigger_signals=True):
+                'Cannot delete form without leaving a tombstone except during testing, domain deletion or '
+                'when deleting system forms')
+        if leave_tombstone:
+            form = self.get(self._id)
+            create_deleted_sql_doc(form.form_id, 'form_processor.XFormInstance', form.domain, form.deleted_on)
```
```suggestion
            create_deleted_sql_doc(self.form_id, 'form_processor.XFormInstance', self.domain, self.deleted_on)
```

millerdev commented on corehq/apps/cleanup/models.py:
```diff
@@ -25,3 +25,13 @@ class Meta:
         ]
+
+
+def create_deleted_sql_doc(doc_id, class_path, domain, deleted_on, deleted_by=None):
```
I think `deleted_by` should be required. If there are system actions with no obvious username associated maybe they can use a name like 'system'?

minhaminha replied:
Currently, we don't keep track of who deleted or even last modified forms and cases (there is a last `modified_by` field for cases but that's only updated on followup form submissions) which means all deletions will basically have the same `deleted_by`. I included the column there to encourage future work so we can either add that field or for cases, extend the usage of `modified_by`. Also, I don't think we need to create tombstones for deleted system forms.

millerdev commented on corehq/apps/cleanup/tests/util.py:
```diff
@@ -47,3 +50,8 @@ def get_cleaned_couch_attrs(cls, couch_model_cls):
+
+@unit_testing_only
+def delete_all_deleted_sql_docs():
+    DeletedSQLDoc.objects.all().delete()
```
This utility should not be necessary, assuming `DeletedSQLDoc` is in pgmain. SQL tests use transactions, which are rolled back to automatically clean up any objects created by the test.

millerdev commented on corehq/form_processor/models/cases.py:
```diff
@@ -294,6 +295,29 @@ def hard_delete_cases(self, domain, case_ids, *, publish_changes=True):
+                deleted_counts = {class_path: queryset.count()}
+                counts.update(deleted_counts)
+            else:
+                for case in queryset:
+                    create_deleted_sql_doc(case.case_id, class_path, case.domain, case.deleted_on)
```
Would it be possible to do this using `DeletedSQLDoc.objects.bulk_create()`? That would be much more efficient than creating them one at a time.

millerdev commented on corehq/form_processor/models/forms.py:
```diff
@@ -423,6 +405,29 @@ def hard_delete_forms(self, domain, form_ids, delete_attachments=True, *, publis
+                deleted_counts = {class_path: queryset.count()}
+                counts.update(deleted_counts)
+            else:
+                for form in queryset:
+                    create_deleted_sql_doc(form.form_id, class_path, form.domain, form.deleted_on)
```
Same recommendation to use `bulk_create`

millerdev reviewed (commented):
Can `DeletedSQLDoc` be found in `raw_doc`? It might be nice to add it (should be pretty easy) if not so we can quickly tell if a form or case has been deleted by looking it up there.

minhaminha committed:
Switch to bulk_create + remove delete_all_deleted_sql_docs

minhaminha committed:
Make sure forms and cases create tombstones on delete when specified, errors when trying to do it outside of tests

minhaminha committed:
Add DeletedSQLDoc table to raw_doc

minhaminha commented on corehq/form_processor/models/cases.py:
```diff
@@ -294,6 +295,29 @@ def hard_delete_cases(self, domain, case_ids, *, publish_changes=True):
+            else:
+                for case in queryset:
+                    create_deleted_sql_doc(case.case_id, class_path, case.domain, case.deleted_on)
+                deleted_counts = queryset.delete()[1]
```
`queryset.delete()` returns a tuple of the total count of deleted docs and a dictionary breaking down those counts by specific object types deleted, which is what we want when logging this info.

millerdev commented on corehq/form_processor/models/forms.py:
```diff
@@ -718,6 +727,15 @@ def unarchive(self, user_id=None, trigger_signals=True):
         if self.is_archived:
             type(self).objects.do_archive(self, False, user_id, trigger_signals)
 
+    def delete(self, leave_tombstone=False):
```
Same as cases, default to `leave_tombstone=True`?

minhaminha replied:
Just made the change to make leaving a tombstone the default but I wonder if there are existing places where system cases/forms are deleted individually? Nothing really turned up when I looked and I assume most of these places probably use bulk deletion anyways but do you have any ideas about whether these places might exist and if I should specify `leave_tombstone=False` to try and keep DeletedSQLDocs (at least for forms and cases) a log only for user driven deletion?

millerdev replied:
> do you have any ideas about whether these places might exist

I think we always want to leave a tombstone if a form or case is being deleted when the domain is not being deleted.

> [should I] specify `leave_tombstone=False` to try and keep DeletedSQLDocs (at least for forms and cases) a log only for user driven deletion?

Interesting. I suppose it would make sense to specify `leave_tombstone=False` in places where system cases are automatically deleted by the system. For example, @gherceg found a place in CommTrack where something like  a _user owner mapping case_ was being deleted prior to linking a new location. Graham should be able to give more details if you can't find the relevant code.

Aside: what happens to the tombstone if a case is deleted, and then later another case is created with the same case id? Does the tombstone get deleted when the new case is created?

minhaminha replied:
> Aside: what happens to the tombstone if a case is deleted, and then later another case is created with the same case id? Does the tombstone get deleted when the new case is created?

hmm I definitely overlooked this scenario and in the current state, it could be problematic if a user tries to delete the second case with the same case_id later on, since it'll pass on creating it, thinking it already exists. How likely is this scenario? Would it help to add `deleted_on` to the unique constraint that already exists for doc_id and object_class_path? or maybe just get rid of that constraint altogether?

millerdev replied:
> How likely is this scenario?

I don't know, but I don't think it will be hard to handle, so we should do that (unless it turns out I am wrong and it's hard).

> Would it help to add `deleted_on` to the unique constraint that already exists for `doc_id` and `object_class_path`? or maybe just get rid of that constraint altogether? or maybe just get rid of that constraint altogether?

Removing that constraint is an intriguing idea. It's useful to guarantee data integrity, but maybe we don't care about duplicates? Another option would be to use an "upsert" operation when writing to the table, which would update rather than insert if a matching record already exists.

millerdev reviewed (commented):
This is looking very good. I think the question about the default is the only thing remaining before this is good to go.

gherceg commented on corehq/form_processor/tests/test_forms.py:
```diff
@@ -428,6 +441,30 @@ def test_nothing_is_deleted_if_dry_run_is_true(self):
+
+        self.assertEqual(DeletedSQLDoc.objects.all().count(), counts['form_processor.XFormInstance'])
+
+    def test_tombstone_is_not_create_if_deleted_on_is_null(self):
```
nit
```suggestion
    def test_tombstone_is_not_created_if_deleted_on_is_null(self):
```

gherceg commented on corehq/form_processor/tests/test_forms.py:
```diff
@@ -428,6 +441,30 @@ def test_nothing_is_deleted_if_dry_run_is_true(self):
+
+        XFormInstance.objects.hard_delete_forms_before_cutoff(self.cutoff, dry_run=False)
+
+        self.assertFalse(list(DeletedSQLDoc.objects.all()))
```
super nit (I just think it is more readable)
```suggestion
        self.assertEqual(DeletedSQLDoc.objects.count(), 0)
```

gherceg commented on corehq/form_processor/models/forms.py:
```diff
@@ -423,6 +406,32 @@ def hard_delete_forms(self, domain, form_ids, delete_attachments=True, *, publis
+                form_tombstones = [DeletedSQLDoc(doc_id=form.form_id, object_class_path=class_path,
+                                                 domain=form.domain, deleted_on=form.deleted_on)
+                                   for form in queryset]
+                for chunk in chunked(form_tombstones, 1000, list):
+                    DeletedSQLDoc.objects.bulk_create(chunk, ignore_conflicts=True)
+                deleted_counts = queryset.delete()[1]
```
Similar to how you override the CommCareCase.delete method to ensure tombstones are created, you could do the same thing on the CommCareCaseManager's delete method for bulk deletes. What do you think?

minhaminha replied:
ohhh cool I thought you couldn't override the delete function for queryset.delete but I guess you can. The one hesitation I have about this is that I'd have exempt all existing processes for domain deletion and for deleting system cases/forms (like the location cases) from creating tombstones. Though maybe something as simple as passing `is_domain_or_system_deletion=True` would be enough (after I find them all).

minhaminha committed:
Make leaving a tombstone a deafult for deleting single forms and cases

gherceg commented on corehq/apps/cleanup/tasks.py:
```diff
@@ -16,34 +16,33 @@
-# @periodic_task(run_every=crontab(minute=0, hour=0), queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery'))
-def permanently_delete_eligible_data(dry_run=False):
+@periodic_task(run_every=crontab(minute=0, hour=0), queue=getattr(settings, 'CELERY_PERIODIC_QUEUE', 'celery'))
+def permanently_delete_and_tombstone_eligible_data(dry_run=False):
```
I have mixed feelings about the name change. Not that I loved the previous name, I just don't think we need to mention the tombstone part in the method name. Especially if this grows into the task that is used to kick off _all_ periodic deletions, in which case it isn't guaranteed to tombstone every object it deletes. It seems like we are heading in a direction where some things are tombstoned, and some things aren't, and it is up to the specific models to decide if they want to tombstone on deletion. Tombstoning has no impact on this method otherwise.

gherceg reviewed (commented):
Apologies for the delayed review and all of the nits 😅

minhaminha commented on corehq/apps/cleanup/models.py:
```diff
@@ -25,3 +25,13 @@ class Meta:
+        with transaction.atomic():
+            DeletedSQLDoc.objects.create(doc_id=doc_id, object_class_path=class_path, domain=domain,
+                                         deleted_on=deleted_on, deleted_by=deleted_by)
+    except IntegrityError:
+        # DeletedSQLDoc already exists for this doc
+        pass
```
Yea this is likely never going to raise that error (unless I guess like if you try to delete the same form multiple times in shell or something). What do you mean by soft asserting here?

millerdev commented on corehq/form_processor/models/cases.py:
```diff
@@ -838,13 +839,14 @@ def save(self, *args, with_tracked_models=False, **kw):
+                raise ValueError(
+                    'Cannot delete case without leaving a tombstone except during testing, domain deletion or '
+                    'when deleting system cases')
+            if leave_tombstone:
+                create_deleted_sql_doc(self.case_id, 'form_processor.CommCareCase', self.domain, self.deleted_on)
```
Given the above condition, `if leave_tombstone` will always evaluate to true, right?
```suggestion
            create_deleted_sql_doc(self.case_id, 'form_processor.CommCareCase', self.domain, self.deleted_on)
```

gherceg replied:
Ha this threw me for a loop. Is it the case that having `leave_tombstone` as a parameter was only necessary for Domains because the tombstone was still a Domain-Deleted couch doc that could "get in the way" of tests? Whereas now that the tombstone is just a separate SQL table, it can't negatively impact tests so we can just always create a tombstone when calling `case.delete()`?

Have we thought through what hard deleting cases will look like? The simple case is when a case has been deleted and a tombstone exists, so now we just have to delete the tombstone, but what if a domain deletion is requested on not yet deleted cases? Are we back to a two step process where we call `case.delete()`, a tombstone is created, and then we have to also delete that tombstone? That seems like where it would be nice be able to set `leave_tombstone` to False right? Which is impossible with the existing code?

minhaminha replied:
> Whereas now that the tombstone is just a separate SQL table, it can't negatively impact tests so we can just always create a tombstone when calling case.delete()?

Case.delete() can't negatively impact tests because there's an explicit check that excludes calls from tests from making tombstones. 

> Are we back to a two step process where we call case.delete(), a tombstone is created, and then we have to also delete that tombstone?

Domain deletion handles case deletion through bulk deletion, which doesn't touch `CommCareCase.delete()` and so currently can't create tombstones. I'm a little confused on how to move forward here because it seems that both you and Daniel agree that it would be best to not create tombstones for case/form deletion via domain deletion and system deletion (specifically user owner mapping cases but there might be other examples). I still need to dig around some more but I'm pretty sure all of these processes already only use bulk delete, and so won't have a chance to create them anyway. My instinct then is to not create a CommCareCaseManager.delete / bulk delete override that makes tombstoning the default and then exclude these processes. I think it might be better to be explicit about when bulk deletion should create tombstones to keep the DeletedSQLDoc table less cluttered with docs we'll probably never reference (unless it we potentially will?).

gherceg replied:
> Case.delete() can't negatively impact tests because there's an explicit check that excludes calls from tests from making tombstones.

Yeah I understand that the check to exclude creating tombstones in tests exists, but I'm implying that we could potentially remove that check since these tombstones (DeletedSQLDoc objects) shouldn't actually cause issues in tests the way Domain-Deleted objects might have, which is especially true since SQL updates are automatically rolled back at the end of tests. Is there a scenario where it makes sense to actually use the `case.delete(leave_tombstone=False)` behavior in tests? Otherwise I think we can remove it and simplify this method.

As for the system forms/cases, it is true that I don't think we should create tombstones for the user owner mapping cases, but I'm not sure I'd extend that logic to other system forms. My reasoning is that the user owner mapping cases have a very specific workflow where cases are frequently hard deleted, and then re-created with the same case id (intentionally). As far as I know, there aren't any other forms/cases that behave this way. And fwiw, you can view the logic for deleting these cases which starts [here](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/apps/users/models.py#L2201-L2208) and [ends up](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/form_processor/models/cases.py#L280-L295) using the [hard_delete_cases](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/sql_accessors/sql_templates/hard_delete_cases_2.sql) SQL function which I assume bypasses the tombstoning logic regardless of if it is defined only on CommCareCase.delete or on CommCareCaseManager.delete as well. 

Could be easiest to offline more about this in slack as well just to make sure we are all on the same page. Personally, I don't feel confident about what the right approach is here.

minhaminha replied:
> Is there a scenario where it makes sense to actually use the case.delete(leave_tombstone=False) behavior in tests? Otherwise I think we can remove it and simplify this method.

Not specifically (except the test I wrote to test that scenario) but I don't think creating tombstones on delete would be relevant to a lot of these tests either and at worst, I feel like all these sql updates could add up to really slow down testing as a whole.

> SQL function which I assume bypasses the tombstoning logic regardless of if it is defined only on CommCareCase.delete or on CommCareCaseManager.delete as well.

thanks for pointing this out for me!! you're right about the this bypassing tombstone creation for cases but it does look like it also eventually calls [hard_delete_cases_and_forms](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/form_processor/backends/sql/processor.py#L63) > `hard_delete_forms` which does make use of the XFormInstanceManager.delete function so we'll have to exclude it explicitly there.

and sorry for the late response but after some more thinking I'm walking back on my last statement about not making tombstoning a default for bulk deletion, since for future work that makes use of it, it's probably better to err on the side of creating tombstones we don't really need than not having them. I'd be very open to offline about this later or do a short call though just to confirm what the next step should be.

millerdev replied:
I think this thread is resolved now. @gherceg do you agree?

gherceg replied:
Agreed

millerdev commented on corehq/form_processor/models/forms.py:
```diff
@@ -727,13 +728,15 @@ def unarchive(self, user_id=None, trigger_signals=True):
+                raise ValueError(
+                    'Cannot delete form without leaving a tombstone except during testing, domain deletion or '
+                    'when deleting system forms')
+            if leave_tombstone:
+                create_deleted_sql_doc(self.form_id, 'form_processor.XFormInstance', self.domain, self.deleted_on)
```
```suggestion
            create_deleted_sql_doc(self.form_id, 'form_processor.XFormInstance', self.domain, self.deleted_on)
```

minhaminha commented:
So after a more thorough look around, it looks like bulk deletion is truly limited to domain deletion (delete_domain.py), the hard_delete_forms_and_cases_in_domain.py command and for deleting user location mapping cases (safe_hard_delete). Bulk deletion is also never actually called for cases - we always use the hard_delete_cases sql function and XFormInstanceManager.hard_delete_forms is the only place we call bulk delete on a form queryset. These are all outside of tests, which I'll continue to not make tombstones for. I don't fully recall what the conclusion was from the conversation on Thursday but in his scenario, it'd be pretty easy still to have bulk deletion default to creating tombstones with just these 3 places excluded (though I still don't fully know under what context hard_delete_forms_and_cases_in_domain is used in so maybe not for that). @millerdev @gherceg

minhaminha committed:
Minor changes to tests (mostly name changes)

+ simplify delete functions

minhaminha committed:
Update tombstone in the unlikely chance of collisions

minhaminha committed:
Add reminders to hard_delete_forms/cases to leave tombstone if used for anything other it’s current usage

So that's anything other than domain deletion, user location mapping case deletion and tests.

millerdev commented on corehq/form_processor/models/cases.py:
```diff
@@ -279,15 +279,27 @@ def soft_undelete_cases(self, domain, case_ids):
         return undeleted_count
 
-    def hard_delete_cases(self, domain, case_ids, *, publish_changes=True):
+    def hard_delete_cases(self, domain, case_ids, *, publish_changes=True, leave_tombstone=False):
```
I think we should set the default to `True` and change all the places that currently use the function to explicitly pass `leave_tombstone=False`. That way a new usage that forgets to pass this argument will quickly find out that tombstones need to be considered.

millerdev commented on corehq/form_processor/models/forms.py:
```diff
@@ -372,15 +372,27 @@ def soft_undelete_forms(self, domain, form_ids):
 
-    def hard_delete_forms(self, domain, form_ids, delete_attachments=True, *, publish_changes=True):
+    def hard_delete_forms(self, domain, form_ids, delete_attachments=True, *,
+                          publish_changes=True, leave_tombstone=False):
```
Same as previous comment, I think this should default to `True`.

minhaminha committed:
Require a leave_tombstone value outside of tests + more tests around that

minhaminha commented:
> Another option would be to use an "upsert" operation when writing to the table, which would update rather than insert if a matching record already exists.

I did actually change the single form and case deletion workflow to use `update_or_create` based off the conversation we had 2 weeks ago but just now realized I need to apply that same logic for the various uses of `bulk_create` too.

minhaminha committed:
Small nit test fix

minhaminha commented:
apologies for the delay but it looks like we don't yet have access to bulk update and create type function in our current django version. As of right now, or bulk creation, if there is a unique key violation, it'll just skip adding that row but once we upgrade to [Django 4.2](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#django.db.models.query.QuerySet.bulk_create) we can have it overwrite/update the row.

minhaminha commented:
@millerdev pinging you just in case you missed this. Is there anything else left to do here?

millerdev commented:
Looks good. One question:

> As of right now, or bulk creation, if there is a unique constraint violation, it'll just skip adding that row

Is there a test for this scenario? I'm concerned that the entire `bulk_create()` call may fail, meaning that multiple tombstones will not be created if any one of them is a duplicate, which sounds like something we do not want.

minhaminha commented:
@millerdev currently the bulk_create function has `ignore_conflicts=True`, which according to the [docs](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#bulk-create) will have it skip over just the failed rows. There's no tests specifically for that but I can add one for both forms and cases hard deletion.

minhaminha committed:
Merge branch 'master' into ml/create-tombstones-cases-forms

minhaminha committed:
Add tests to cover ignore_conflicts behavior in bulk_create

minhaminha committed:
Return more accurate deletion count

minhaminha committed:
Add additional logging for tombstones created + fix and add tests around that

minhaminha committed:
Delete forms and cases in chunks to be less memory intensive

millerdev commented on corehq/apps/cleanup/tasks.py:
```diff
@@ -40,9 +40,15 @@ def permanently_delete_eligible_data(dry_run=False):
     for table, count in form_counts.items():
-        logger.info(f"{dry_run_tag}{count} {table} objects were deleted.")
+        if table == 'tombstone':
+            logger.info(f"{count} XFormInstance {table}s were created.")
```
Why is `{dry_run_tag}` not used here?

minhaminha replied:
tombstones don't need to and won't be created for dry runs - that's just to get a sense of how many forms/cases will be deleted during the next run

millerdev commented on corehq/apps/cleanup/tasks.py:
```diff
@@ -40,9 +40,15 @@ def permanently_delete_eligible_data(dry_run=False):
     for table, count in case_counts.items():
-        logger.info(f"{dry_run_tag}{count} {table} objects were deleted.")
+        if table == 'tombstone':
+            logger.info(f"{count} CommCareCase {table}s were created.")
```
same as above

millerdev commented on corehq/form_processor/models/cases.py:
```diff
@@ -334,11 +334,13 @@ def hard_delete_cases_before_cutoff(self, cutoff, dry_run=True):
                 case_tombstones = [DeletedSQLDoc(doc_id=case.case_id, object_class_path=class_path,
                                                  domain=case.domain, deleted_on=case.deleted_on)
                                    for case in queryset]
```
Iterating over `queryset` (which loads the full result set into memory) is the thing that would significant memory pressure, not the `.delete()` call below. A possibly more efficient technique would be to run the query in a loop and limit the number of results. Something like this
```py
while cases := queryset[:2000]:
    chunk = [DeletedSQLDoc(doc_id=case.case_id, object_class_path=class_path,
                             domain=case.domain, deleted_on=case.deleted_on)
             for case in cases]
    tombstone_count = len(DeletedSQLDoc.objects.bulk_create(chunk, ignore_conflicts=True))
    counts['tombstone'] += tombstone_count
    chunked_queryset = self.using(db_name).filter(case_id__in=[doc.doc_id for doc in chunk])
    deleted_counts = chunked_queryset.delete()[1]
    for obj_class, count in deleted_counts.items():
        counts[obj_class] += count
```
Aside: can't believe I'm recommending the walrus operator, but I think this is the one place where it's most justified.

minhaminha replied:
thanks for the suggestion! Just curious though, the django docs has this line:
> You cannot call delete() on a [QuerySet](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet) that has had a slice taken or can otherwise no longer be filtered.

Does this apply here to the chunked queryset? or am I just misunderstanding this line?

minhaminha replied:
ohhh nevermind I misread the code change

minhaminha committed:
Apply suggestion to chunk queryset instead of tombstone list
